File: java/225_Implement_Stack_using_Queues.java
/*
https://leetcode.com/problems/implement-stack-using-queues/
Implement the following operations of a stack using queues.
    push(x) -- Push element x onto stack.
    pop() -- Removes the element on top of the stack.
    top() -- Get the top element.
    empty() -- Return whether the stack is empty.

When pushing, cycle round all previous items (pop then push) so newest is always at front of queue.
Time - O(n) to push, O(1) to pop and peek.
Space - O(n)
*/

class MyStack {

    private Queue<Integer> queue = new LinkedList<>();

    // Push element x onto stack.
    public void push(int x) {
        int length = queue.size();
        queue.add(x);
        for (int i = 0; i < length; ++i) {
            queue.add(queue.poll());
        }
    }

    // Removes the element on top of the stack.
    public void pop() {
        queue.poll();
    }

    // Get the top element.
    public int top() {
        return queue.peek();
    }

    // Return whether the stack is empty.
    public boolean empty() {
        return queue.isEmpty();
    }
}

File: python_1_to_1000/225_Implement_Stack_Using_Queues.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/implement-stack-using-queues/
# Implement the following operations of a stack using queues.
# push(x) -- Push element x onto stack.
# pop() -- Removes the element on top of the stack.
# top() -- Get the top element.
# empty() -- Return whether the stack is empty.
# Notes:
# You must use only standard operations of a queue -- which means only push to back, peek/pop from front, size,
# and is empty operations are valid.
# Depending on your language, queue may not be supported natively. You may simulate a queue by using a list or
# deque (double-ended queue), as long as you use only standard operations of a queue.
# You may assume that all operations are valid (for example, no pop or top operations will be called on an empty stack).

# To insert, append to deque. To pop, move ell elements to new_queue until there is only one element and return that.
# To top, similar to pop except all elements are moved to new_queue. After pop or top, set queue to new_queue.
# Time - O(1) for push and empty, O(n) for pop and top.
# Space - O(n)

from collections import deque

class MyStack(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.queue = deque()

    def push(self, x):
        """
        Push element x onto stack.
        :type x: int
        :rtype: void
        """
        self.queue.appendleft(x)

    def pop(self):
        """
        Removes the element on top of the stack and returns that element.
        :rtype: int
        """
        new_queue = deque()
        while True:
            x = self.queue.pop()
            if not self.queue:
                self.queue = new_queue
                return x
            new_queue.appendleft(x)

    def top(self):
        """
        Get the top element.
        :rtype: int
        """
        new_queue = deque()
        while self.queue:
            x = self.queue.pop()
            new_queue.appendleft(x)
        self.queue = new_queue
        return x

    def empty(self):
        """
        Returns whether the stack is empty.
        :rtype: bool
        """
        return len(self.queue) == 0